### ISSUE TITLE
Visualization in Pyro

### ISSUE THREAD

**Issue by eb8680** [uber/pyro#20]
Visualization is critical for understanding and debugging models and inference algorithms.  What are our visualization desiderata and what features are necessary to support them?  

It seems to me that eventually, we'll need support for automatic visualization of samples, scores, and summary statistics and their evolution in execution traces, trace and autograd graph structures, losses and individual loss terms and their evolution in inference algorithms, NaN detection, and parameter/parameter gradient summary statistics (variance, norm, etc).  We may also want custom visualizations for specific primitive distributions, and for our model criticism stuff.

Given that we've settled on visdom as our primary visualization tool, what Pyro-specific things do we need to implement, and what sort of interface(s) should we provide?  Should we evaluate other tools as well?

Some references:
- [webppl-viz](https://github.com/probmods/webppl-viz)
- [visdom](https://github.com/facebookresearch/visdom)
- [TensorBoard](https://github.com/tensorflow/tensorflow/blob/r1.2/tensorflow/tensorboard/README.md)
- [TensorBoard and Edward](http://edwardlib.org/tutorials/tensorboard)
- TensorBoard decoupled from the rest of TF: [crayon](https://github.com/torrvision/crayon) and [dmlc's fork](https://github.com/dmlc/tensorboard)
- [TensorBoard for PyTorch](https://github.com/lanpa/tensorboard-pytorch)
- [Visualization in PyMC](http://pymc-devs.github.io/pymc3/api/plots.html)

**Comment by eb8680** [comment]
Closing in favor of more targeted discussion in future
